cpp/src/decisiontree/levelalgo/levelfunc_classifier.cuh

```diff
@@ -0,0 +1,136 @@
+  const ML::cumlHandle_impl& handle, const T* data, const int* labels,
+  unsigned int* rowids, const std::vector<unsigned int>& feature_selector,
+  int n_sampled_rows, const int nrows, const int n_unique_labels,
+  const int nbins, const int maxdepth, const int maxleaves,
+  const int min_rows_per_node, const ML::CRITERION split_cr, int& depth_cnt,
+  int& leaf_cnt, std::shared_ptr<TemporaryMemory<T, int>> tempmem) {
+  const int ncols = feature_selector.size();
+  MLCommon::updateDevice(tempmem->d_colids->data(), feature_selector.data(),
+                         feature_selector.size(), tempmem->stream);
+
+  unsigned int* flagsptr = tempmem->d_flags->data();
+  unsigned int* sample_cnt = tempmem->d_sample_cnt->data();
+  setup_sampling(flagsptr, sample_cnt, rowids, nrows, n_sampled_rows,
+                 tempmem->stream);
+  std::vector<int> histvec(n_unique_labels, 0);
+  T initial_metric;
+  if (split_cr == ML::CRITERION::GINI) {
+    initial_metric_classification<T, GiniFunctor>(labels, sample_cnt, nrows,
+                                                  n_unique_labels, histvec,
+                                                  initial_metric, tempmem);
+  } else {
+    initial_metric_classification<T, EntropyFunctor>(labels, sample_cnt, nrows,
+                                                     n_unique_labels, histvec,
+                                                     initial_metric, tempmem);
+  }
+  size_t total_nodes = 0;
+  for (int i = 0; i <= maxdepth; i++) {
+    total_nodes += pow(2, i);
+  }
+
+  std::vector<std::vector<int>> sparse_histstate;
+  sparse_histstate.resize(total_nodes, std::vector<int>(n_unique_labels));
+  sparse_histstate[0] = histvec;
+
+  std::vector<SparseTreeNode<T, int>> sparsetree;
+  sparsetree.reserve(total_nodes);
+  SparseTreeNode<T, int> sparsenode;
+  sparsenode.best_metric_val = initial_metric;
+  sparsetree.push_back(sparsenode);
+  int sparsesize = 0;
+  int sparsesize_nextitr = 0;
+
+  int n_nodes = 1;
+  int n_nodes_nextitr = 1;
+  std::vector<int> sparse_nodelist;
+  sparse_nodelist.reserve(pow(2, maxdepth));
+  sparse_nodelist.push_back(0);
+  //this can be depth loop
+
+  //Setup pointers
+  unsigned int* d_histogram = tempmem->d_histogram->data();
+  unsigned int* h_histogram = tempmem->h_histogram->data();
+  int* h_split_binidx = tempmem->h_split_binidx->data();
+  int* d_split_binidx = tempmem->d_split_binidx->data();
+  int* h_split_colidx = tempmem->h_split_colidx->data();
+  int* d_split_colidx = tempmem->d_split_colidx->data();
+  unsigned int* h_new_node_flags = tempmem->h_new_node_flags->data();
+  unsigned int* d_new_node_flags = tempmem->d_new_node_flags->data();
+  unsigned int* d_colids = tempmem->d_colids->data();
+
+  for (int depth = 0; (depth < maxdepth) && (n_nodes_nextitr != 0); depth++) {
+    depth_cnt = depth + 1;
+    n_nodes = n_nodes_nextitr;
+    sparsesize = sparsesize_nextitr;
+    sparsesize_nextitr = sparsetree.size();
+    ASSERT(
```

Comment:
I'm guessing this ASSERT is here because if somehow there are more nodes at a level than was told to us by the user, it can potentially cause OOM?

Reply:
Yeah. Right now I have made a static max_depth of **'20'** ; however user can still go deeper than 20 but only if the number of nodes (split nodes not leaves) at a particular level are less than the max which is pow(2, 20) . This is possible because we operate on sparse nodes. Hence the assert is dataset dependent here. 

Ideally I would like to query the available memory on the GPU to make this decision.

Reply:
> Ideally I would like to query the available memory on the GPU to make this decision. 

It will be great when this is possible. I think most algorithms in cuML will benefit from this.

Reply:
But there will still be a hard limit of 32 depth, given the current impl., even if there was sufficient memory, right? That is according to the comment in `split_level_kernel` in `/decisiontree/levelalgo/common_kernel.cuh`. It might be helpful to clarify that somewhere or maybe default max_depth = -1 to 32 for the time being.

Reply:
@myrtw  May be an assert earlier for Quantile algo. ??

Reply:
I see. I had misunderstood the 32 constraint. Thanks for clarifying!